moztelemetry/dataset.py

```diff
@@ -167,8 +194,7 @@ def select(self, *properties, **aliased_properties):
         new_selection = self.selection.copy()
         new_selection.update(merged_properties)
 
-        return Dataset(self.bucket, self.schema, store=self.store, prefix=self.prefix,
-                       clauses=self.clauses, selection=new_selection)
+        return self._copy(selection=new_selection)
```

Comment:
Yes, that would work, but it would change the semantics of the Dataset class and potentially break existing use cases.

From the Dataset docstring:

> The result of refining a Dataset is a Dataset itself

I understand the contract to be that Dataset instances are effectively immutable and that methods like `select` and `where` should never mutate the object's state, but rather return a new object representing the new state.

Introducing `_copy` is intended as a refactoring to make creating a "refined" Dataset safer, rather than relying on `select` and `where` to each correctly pass all the correct params to the `Dataset` contstructor, which is what they previously were doing.

Reply:
Technically, `type(self) == Dataset` :). But I see how it would change the semantics of the class:

```python
x = Dataset(...)
y = x.where(...).select(...)
z = x.where(...)
```

Using self would lead to counter-intuitive behavior.